### ISSUE TITLE
human_like_ai_suggestions

### ISSUE THREAD

**Issue by venylol** [Nyanyan/Egaroucid#596]
# Proposal: Simulate Human-Like Play

This proposal is my own original idea. The concrete mathematical formulas and
algorithmic details below were suggested by ChatGPT after discussing the idea.

I am opening this as a possible direction for the existing "Human-like AI with
NN" issue. I do not think this has to replace the NN idea completely. My main
point is that a data-driven, non-searching human-like AI may need a stronger
statistical baseline before a neural network policy becomes reliable.

## 1. Why I Think Training a Neural Network Directly on "Human Moves" Is Hard

Training a human-like policy is different from distilling a strong engine.

When we distill a high-level engine, the target is relatively stable: for a
given position, the best move or best evaluation target is close to a
well-defined answer.

Human moves are much noisier:

- Different players have very different strengths.
- The same position may have many human moves with very different quality.
- A "human" move distribution is not necessarily a "good" move distribution.
- If we train on games from many players, the target policy mixes many skill
  levels and styles together.
- If we train on only one player, the style is clearer, but the dataset becomes
  too small and too narrow.

So a raw NN policy trained from human games may learn a mixture of opening
habits, mistakes, style bias, and player-strength noise. That does not mean the
NN approach is impossible, but I think it needs good metadata and calibration
targets first.

My suggestion is to separate the problem into two parts:

1. Human opening choice probability.
2. Human loss tendency by game phase.

These two parts can be measured directly from game records and then used as a
baseline, a control target, or training metadata for a future NN.

### 2.1 Opening Phase: Human Move Frequency

For the opening phase, I suggest using real human move frequencies.

The process is:

1. Read human game records.
2. Convert each position to a normalized board key.
3. Normalize by side-to-move perspective and board symmetries.
4. Count how often each normalized opening node appears.
5. At runtime, enumerate legal child positions from the current position.
6. Use the retained child nodes' occurrence counts as the human move frequency.

For a normalized position:

```text
P_human(move_i)
  = occurrence_count(child_position_i)
    / sum_j occurrence_count(child_position_j)
```

So the packaged prototype does not store a direct parent-to-move probability
table. It stores normalized board nodes and legal retained child edges, then
derives the move distribution from child-node occurrence counts. This gives a
simple and explainable human opening model. It can reproduce both common
openings and less common human choices.

The current experimental sampling idea is controlled by a `human loss level`.
For the opening phase, the level does not mean engine stone loss directly. It
means how far the sampled opening move is from the most common human move.

Let:

```text
best_count  = max count among legal in-book human moves
worst_count = min count among legal in-book human moves
level       = user setting, for example 0..14
target      = level / 14
```

For each move:

```text
human_frequency_loss(move)
  = (best_count - count(move)) / (best_count - worst_count)
```

So:

- `0.0` means the most common human move.
- `1.0` means the rarest human move among the available human moves.

Then sample by an exponential target function:

```text
human_mass(move) = count(move) / sum_j count(move_j)

weight(move)
  = human_mass(move)^frequency_exponent
    * exp(-focus * abs(human_frequency_loss(move) - target))
    * tail_boost(move)
```

In the current Sonetto prototype:

```text
focus = 8 - 4 * target
frequency_exponent = max(0.2, 1 - 0.75 * target)
tail_boost(move) = (0.05 + human_frequency_loss(move)) ^ (0.85 * target)
```

Finally:

```text
P_sample(move) = weight(move) / sum_j weight(move_j)
```

This keeps the behavior human-frequency-based, while allowing the user to
choose whether the AI should prefer common human moves or more offbeat human
moves. In the current prototype, level 0 falls back to the most popular retained
human book move, and the maximum level falls back to the most offbeat retained
human book move.

### 2.2 Middle Game and Endgame: Human Loss Curve

For the middle game and endgame, I suggest measuring how much value humans lose
at different phases of the game.

For each human move in the dataset:

1. Evaluate the best move with Egaroucid.
2. Evaluate the move actually played by the human.
3. Define human stone loss as:

```text
raw_stone_loss = best_score - human_score
clipped_stone_loss = max(0, raw_stone_loss)
```

Then aggregate these losses by ply or by phase. In my current experiment I used
5-ply chunks:

```text
chunk_index = floor((global_ply - 1) / 5) + 1
```

The current THOR-based metadata I computed is:

| chunk | ply range | empties before | sample count | avg raw loss | avg clipped loss | avg tanh40 clipped loss |
|---:|---:|---:|---:|---:|---:|---:|
| 1 | 1-5 | 60-56 | 687740 | 0.278143 | 0.279472 | 0.277005 |
| 2 | 6-10 | 55-51 | 687739 | 0.712500 | 0.771287 | 0.758272 |
| 3 | 11-15 | 50-46 | 687713 | 0.896124 | 1.066257 | 1.024783 |
| 4 | 16-20 | 45-41 | 687668 | 0.991111 | 1.261287 | 1.173566 |
| 5 | 21-25 | 40-36 | 687648 | 1.083972 | 1.397718 | 1.252834 |
| 6 | 26-30 | 35-31 | 687639 | 1.161522 | 1.518986 | 1.311180 |
| 7 | 31-35 | 30-26 | 687615 | 1.906613 | 2.127330 | 1.791043 |
| 8 | 36-40 | 25-21 | 687499 | 2.389429 | 2.466203 | 2.012100 |
| 9 | 41-45 | 20-16 | 687003 | 2.413279 | 2.414036 | 1.921211 |
| 10 | 46-50 | 15-11 | 685334 | 1.917159 | 1.919181 | 1.492041 |
| 11 | 51-55 | 10-6 | 683090 | 1.113792 | 1.114377 | 0.842997 |
| 12 | 56-60 | 5-1 | 674511 | 0.260031 | 0.260052 | 0.192389 |

There is also a fitted continuous phase curve currently used by my local
prototype frontend:

```text
k(e) = exp(b3*x^3 + b2*x^2 + b1*x + b0)
x = (empties_before - 38) / 22

b3 = -1.2785893008416975
b2 = -1.316874434993233
b1 =  0.20756050085814595
b0 = -0.0002464906829407083
```

This curve was fitted from chunk 3 to chunk 12, corresponding roughly to
ply 11-60.

The attached Python runtime also contains older fallback/default model constants
for offline experiments. The active curve in the included frontend metadata is
the `pum-full-bookfiltered-phase-curve-v2-chunk3-12` curve shown above.

### 2.3 Human Loss Level as an Expected Loss Budget

The user-facing control could be `Human Loss Level`.

Internally, this can be mapped to an expected total stone loss budget for one
game:

```text
B = expected total loss for this game
```

If human opening simulation is enabled, the opening phase is handled by the
human opening frequency model. The actual engine loss caused by those opening
moves can be deducted from the total budget:

```text
B_remaining = B - opening_actual_loss
```

However, opening moves should not be included in the middle/endgame loss-curve
normalization, because the opening model is based on human frequency instead of
engine loss.

After leaving the opening book, distribute the remaining budget over the
remaining own moves using the human loss curve.

For the current move:

```text
c_t = human loss curve coefficient for the current phase
S   = sum of human loss curve coefficients for all remaining own moves

target_loss_t = B_remaining * c_t / S
```

This means phases where humans usually lose more value receive more of the
remaining loss budget.

### 2.4 Middle/End Move Sampling with Softmax

For middle/endgame move choice, evaluate candidate moves and compute their
loss:

```text
loss(move_i) = max(0, best_score - score(move_i))
```

Then sample moves near the current target loss:

```text
weight(move_i)
  = exp(-abs(loss(move_i) - target_loss_t) / temperature)

P_sample(move_i)
  = weight(move_i) / sum_j weight(move_j)
```

This is a softmax-like sampling method. It does not always choose the best move,
but it also does not choose random bad moves. It tries to choose moves whose
loss is close to the human-like expected loss for that phase.

In short:

- Opening phase: use human frequency and an exponential target function.
- Middle/endgame: use engine-evaluated loss and softmax around a target loss.

This keeps the two meanings separate:

- Opening `human loss level` means common vs rare human opening choice.
- Middle/endgame `human loss level` means expected engine-value loss.

## 3. Game Record Data Source

The game record source I am referring to is:

https://drive.google.com/drive/folders/16zeUWeGuaKE8j-0irViE8hozfwsBJlBR

This folder contains data from the database by Michele Borassi, the author of
Othello Sensei. It includes some recent Othello Quest game records and the
complete THOR game record set.

My current computed metadata only uses all data from the THOR part.

For more realistic or higher-quality sampling, the same scripts can be used
again after adapting or converting the Othello Quest records into the same kind
of position/move stream. This may be useful because Othello Quest records can
better reflect modern online human play, while THOR is valuable because it is
complete and historically broad.

## 4. Reproduction Scripts

I attached a zip package containing the relevant reproduction scripts and small
metadata files.

The package contains two reference parts:

### 4.1 Human Opening Scripts

These scripts reproduce the opening-frequency idea:

- Normalize positions.
- Extract THOR opening nodes.
- Count normalized opening-node occurrences.
- Derive runtime move frequencies from retained legal child nodes.
- Map positions to Egaroucid scores.
- Build a compact runtime opening book.

The main scripts are:

```text
sonetto_opening/scripts/extract_thor_opening_nodes.py
sonetto_opening/scripts/map_thor_opening_node_scores.py
sonetto_opening/scripts/sample_thor_opening_node_winrates.py
sonetto_opening/scripts/build_thor_opening_board_nodes_payload.py
sonetto_opening/scripts/build_thor_opening_board_minimax_book.py
sonetto_opening/scripts/build_thor_minimax_value_runtime_book.py
sonetto_opening/scripts/evaluate_egaroucid_baseline.py
```

### 4.2 Human Loss Curve Scripts

These scripts reproduce the human loss curve idea:

- Read THOR records.
- Evaluate human moves with Egaroucid Console.
- Compute `best_score - human_score`.
- Summarize raw, clipped, and `tanh40`-transformed loss by ply/chunk.
- Export model/curve data for runtime use.

The main scripts are:

```text
human_loss_curve/server_bundle/thor_all_rawdata_server.py
human_loss_curve/server_bundle/run_thor_all_rawdata_server.cmd
human_loss_curve/elo_estimator_pipeline/model_runtime.py
human_loss_curve/elo_estimator_pipeline/export_sequence_dataset.py
human_loss_curve/elo_estimator_pipeline/export_raw_only_sequence_dataset.py
human_loss_curve/elo_estimator_pipeline/extract_model_for_typescript.py
human_loss_curve/elo_estimator_pipeline/book_prefilter.py
human_loss_curve/elo_estimator_pipeline/early_offbook_common.py
human_loss_curve/elo_estimator_pipeline/build_opening_prefix_stats.py
human_loss_curve/elo_estimator_pipeline/run_egaroucid_unique_positions.py
human_loss_curve/elo_estimator_pipeline/run_egaroucid_unique_positions_loop.py
```

The zip does not include very large files such as:

- Full WThor source files.
- Egaroucid Console binary/resources.
- The full raw detail CSV, which is about 1.9 GB locally.
- Large intermediate JSON datasets.

It does include small metadata files, summary CSVs, and samples so that the
pipeline structure and current computed results can be inspected.

## Final Note

At the moment, I have only computed the metadata from all THOR data.

If the goal is to make the behavior more realistic for modern human online
play, the Othello Quest records from Michele Borassi's database can be sampled
after adding a small converter/adapter for their record format. The Egaroucid
evaluation, position normalization, loss aggregation, and curve-fitting parts
should still be reusable. That should give a better modern human move
distribution and possibly a better human loss curve.

I hope this can be useful either as:

1. A non-NN baseline for human-like play.
2. A calibration target for a future NN policy.
3. Additional metadata for clustering different human play styles.

[human_like_ai_issue_attachment.zip](https://github.com/user-attachments/files/28808814/human_like_ai_issue_attachment.zip)

**Comment by venylol** [comment]
追記です。

石損設定と human loss level の関係について、少し気になったエッジケースがあります。

ここでは、石損設定は「AI を最終的にどれくらい弱くするか」という主な期待値で、human loss level は「その弱さをどれくらい人間らしい形で出すか」という補助的な設定だと考えるのが自然だと思います。

例えば、低い石損 + 高い human loss level の場合、AI が人間らしい悪手を選びすぎて、設定された石損期待値よりかなり弱くなってしまうのは避けたいです。
この場合は、低い石損期待値に合う範囲で、より人間らしい手や少し珍しい手を選ぶ、という挙動がよさそうです。

逆に、高い石損 + 低い human loss level の場合、AI がほとんど最善手ばかり選んで強くなりすぎるのも避けたいです。
石損設定が高いなら、平均的にはその石損期待値に近づくべきですが、human loss level が低い場合は、あまり人間風・意図的な悪手っぽくならない方法で弱くするのがよいと思います。

つまり、

> 石損設定 = どれくらい弱くするか
> human loss level = どのような形で弱くするか

という関係にすると、既存の石損設定とも自然につながると思います。

**Comment by Nyanyan** [comment]
Thank you for this fantastic proposal. This is an excellent idea!
Here are my thoughts on your suggestions:

- Opening Phase: The idea based on real human opening statistics is highly practical and looks like something we can implement right away.

- Middle/Endgame Phase: In addition to your proposal, I feel it would be even more natural if we could introduce a neural network that judges whether a position is difficult (prone to human error) or simple. Coincidentally, I am planning to develop exactly this kind of network for another project of mine in the near future, so we might be able to leverage or port that model over to this implementation.
- Feature Integration: As suggested in the follow-up discussion, it would be a great approach to add a "Human-like Mode" to the existing AI loss setting. This way, the AI can induce human-like mistakes dynamically based on the specific loss budget or level configured by the user.

by the way, my another project is this othello teaching robot: https://youtu.be/rA0DlM5BObM
